File: algorithms/advanced_data_structures/suffix_tree.py
# Suffix Tree:
# Time Complexity:
# building: O(N)
# search for a substring of length: O(m)
#
# Space Complexity:
# Memory:   O(N)

class SuffixTreeNode:
    def __init__(self, start, end):
        self.start = start
        self.end = end
        self.children = {}
        self.suffix_link = None
        self.suffix_index = -1

class SuffixTree:
    def __init__(self, text):
        self.text = text
        self.root = SuffixTreeNode(-1, -1)
        self.active_node = self.root
        self.active_edge = -1
        self.active_length = 0
        self.remaining_suffix_count = 0
        self.leaf_end = [-1]  # Глобальный конец для всех листьев
        self.root.suffix_link = self.root
        self.size = len(text)

        self.build_suffix_tree()

    def edge_length(self, node):
        if node == self.root:
            return 0

        return node.end[0] - node.start + 1

    def walk_down(self, curr_node):
        length = self.edge_length(curr_node)

        if self.active_length >= length:
            self.active_edge += length
            self.active_length -= length
            self.active_node = curr_node

            return True

        return False

    def extend_suffix_tree(self, pos):
        self.leaf_end[0] = pos  # O(1)
        self.remaining_suffix_count += 1
        last_new_node = None

        while self.remaining_suffix_count > 0:
            if self.active_length == 0:
                self.active_edge = pos

            char = self.text[self.active_edge]

            if char not in self.active_node.children:
                self.active_node.children[char] = SuffixTreeNode(pos, self.leaf_end)
                if last_new_node is not None:
                    last_new_node.suffix_link = self.active_node
                    last_new_node = None
            else:
                next_node = self.active_node.children[char]
                if self.walk_down(next_node):
                    continue

                if self.text[next_node.start + self.active_length] == self.text[pos]:
                    if last_new_node is not None and self.active_node != self.root:
                        last_new_node.suffix_link = self.active_node
                        last_new_node = None
                    self.active_length += 1
                    break

                split_end = [next_node.start + self.active_length - 1]
                split_node = SuffixTreeNode(next_node.start, split_end)
                self.active_node.children[char] = split_node

                split_node.children[self.text[pos]] = SuffixTreeNode(pos, self.leaf_end)
                next_node.start += self.active_length
                split_node.children[self.text[next_node.start]] = next_node

                if last_new_node is not None:
                    last_new_node.suffix_link = split_node
                last_new_node = split_node

            self.remaining_suffix_count -= 1

            if self.active_node == self.root and self.active_length > 0:
                self.active_length -= 1
                self.active_edge = pos - self.remaining_suffix_count + 1
            elif self.active_node != self.root:
                self.active_node = self.active_node.suffix_link

    def set_suffix_index_by_dfs(self, node, label_height):
        if node == self.root:
            pass
        else:
            if not node.children:
                node.suffix_index = self.size - label_height
                return

        for child in node.children.values():
            self.set_suffix_index_by_dfs(child, label_height + self.edge_length(child))

    def build_suffix_tree(self):
        for phase in range(self.size):
            self.extend_suffix_tree(phase)

        self.set_suffix_index_by_dfs(self.root, 0)

# Where to use it:
# 1. Pattern Matching
# 2. Поиск всех индексов вхождений
# 3. Самая длинная повторяющаяся подстрока
# 4. Наибольшая общая подстрока двух строк (LCS)

# How to use it:
# 1. Pattern Matching
# def is_substring(tree, pattern):
#     node = tree.root
#     p_idx = 0
#
#     while p_idx < len(pattern):
#         char = pattern[p_idx]
#         if char not in node.children:
#             return False
#
#         child = node.children[char]
#
#         edge_len = child.end[0] - child.start + 1
#         for i in range(edge_len):
#             if p_idx == len(pattern):
#                 return True
#
#             if tree.text[child.start + i] != pattern[p_idx]:
#                 return False
#             p_idx += 1
#
#         node = child
#
#     return True
#
# 2. Самая длинная повторяющаяся подстрока
# def longest_repeated_substring(node, current_depth, max_depth, max_start_index):
#     if not node.children:
#         return max_depth, max_start_index
#
#     if current_depth > max_depth and current_depth > 0:
#         max_depth = current_depth
#         max_start_index = list(node.children.values())[0].start - current_depth
#
#     for child in node.children.values():
#         edge_length = child.end[0] - child.start + 1
#         max_depth, max_start_index = longest_repeated_substring(
#             child, current_depth + edge_length, max_depth, max_start_index
#         )
#
#     return max_depth, max_start_index
#
#

# Идея как у Suffix Array, но Tree.
# Скорее всего ты это не будишь использовать, намного более часто ты будешь использовать:
# Prefix Function (KMP)
# Z-Function
# Trie
# Hashing
# Suffix Array
# Suffix Automaton

File: cpp_algorithms/advanced_data_structures/suffix_tree.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

// Suffix Tree (Ukkonen)
// Build: O(n)
// Pattern search: O(m)
// Space: O(n)

class SuffixTree {
    struct Node {
        int start;
        shared_ptr<int> end;
        unordered_map<char, int> children;
        int suffixLink = 0;
        int suffixIndex = -1;

        Node(int start, shared_ptr<int> end) : start(start), end(move(end)) {}
    };

    string text;
    vector<Node> tree;

    int root;
    int activeNode;
    int activeEdge = -1;
    int activeLength = 0;
    int remainingSuffixCount = 0;

    shared_ptr<int> leafEnd;
    int n;

    int newNode(int start, shared_ptr<int> end) {
        tree.emplace_back(start, move(end));
        return (int)tree.size() - 1;
    }

    int edgeLength(int node) {
        if (node == root)
            return 0;

        return *tree[node].end - tree[node].start + 1;
    }

    bool walkDown(int node) {
        int len = edgeLength(node);

        if (activeLength >= len) {
            activeEdge += len;
            activeLength -= len;
            activeNode = node;
            return true;
        }

        return false;
    }

    void extend(int pos) {
        *leafEnd = pos;
        remainingSuffixCount++;

        int lastNewNode = -1;

        while (remainingSuffixCount > 0) {
            if (activeLength == 0)
                activeEdge = pos;

            char c = text[activeEdge];

            if (!tree[activeNode].children.count(c)) {
                int leaf = newNode(pos, leafEnd);
                tree[activeNode].children[c] = leaf;

                if (lastNewNode != -1) {
                    tree[lastNewNode].suffixLink = activeNode;
                    lastNewNode = -1;
                }
            } else {
                int next = tree[activeNode].children[c];

                if (walkDown(next))
                    continue;

                if (text[tree[next].start + activeLength] == text[pos]) {
                    if (lastNewNode != -1 && activeNode != root) {
                        tree[lastNewNode].suffixLink = activeNode;
                        lastNewNode = -1;
                    }

                    activeLength++;
                    break;
                }

                auto splitEnd =
                    make_shared<int>(tree[next].start + activeLength - 1);

                int split = newNode(tree[next].start, splitEnd);

                tree[activeNode].children[c] = split;

                int leaf = newNode(pos, leafEnd);
                tree[split].children[text[pos]] = leaf;

                tree[next].start += activeLength;
                tree[split].children[text[tree[next].start]] = next;

                if (lastNewNode != -1)
                    tree[lastNewNode].suffixLink = split;

                lastNewNode = split;
            }

            remainingSuffixCount--;

            if (activeNode == root && activeLength > 0) {
                activeLength--;
                activeEdge = pos - remainingSuffixCount + 1;
            } else if (activeNode != root) {
                activeNode = tree[activeNode].suffixLink;
            }
        }
    }

    void setSuffixIndex(int node, int height) {
        if (tree[node].children.empty()) {
            tree[node].suffixIndex = n - height;
            return;
        }

        for (auto [c, child] : tree[node].children)
            setSuffixIndex(child, height + edgeLength(child));
    }

public:
    SuffixTree(const string& s) : text(s), n(s.size()) {
        leafEnd = make_shared<int>(-1);

        root = newNode(-1, make_shared<int>(-1));
        tree[root].suffixLink = root;

        activeNode = root;

        for (int i = 0; i < n; i++)
            extend(i);

        setSuffixIndex(root, 0);
    }

    bool contains(const string& pattern) {
        int node = root;
        int p = 0;

        while (p < (int)pattern.size()) {
            char c = pattern[p];

            if (!tree[node].children.count(c))
                return false;

            int child = tree[node].children[c];

            for (int i = tree[child].start;
                 i <= *tree[child].end && p < (int)pattern.size();
                 i++, p++) {

                if (text[i] != pattern[p])
                    return false;
            }

            node = child;
        }

        return true;
    }
};
